### DISCUSSION TITLE
How to make your Azure DevOps CI/CD pipeline faster

### DISCUSSION THREAD

**Discussion by giscus** [timdeschryver/timdeschryver.dev#266] [Announcements]
# How to make your Azure DevOps CI/CD pipeline faster

Make your CI/CD pipeline smarter and only build (and release) what's modified to optimize a run

https://timdeschryver.dev/blog/how-to-make-your-azure-devops-ci-cd-pipeline-faster

**Comment by GrumpyMeow** [comment]
Nice to read how implemented your CI-pipeline. I’ve implemented the CI-pipeline for the company I work for. It’s a constant quest for me to further improve on it.
I’ve chosen to use the environment variables NX_HEAD and NX_BASE. 

With non-pull-request builds, I do:
The NX_BASE is set with the commit of the “last successful run of the current pipeline of the current branch”. To fetch this commitid I call the Azure DevOps API. This has proven to be a stable solution for us. 
If a build fails, the next build will ensure that the changes of the previous build will also be taken into account.

A few other optimizations I recently did:
* “sharding” of Storybook by using “—exclude=“*,!tag:shard1” to run matrix on multiple agents.
* caching on Azure Storage of jest- and eslint-cache.

**Reply by GrumpyMeow** [reply]
I forgot one:
* “serve-static” for E2E-tests

**Reply by GrumpyMeow** [reply]
Worst case scenario our pipeline takes 2 to 3 hours 🙈. But that’s because we have 190 projects. Especially Storybook and Cypress take a lot of time for us.

**Reply by timdeschryver** [reply]
Thanks for sharing @GrumpyMeow I especially like the usage of shards!
